lib/containers/runtime.pm

```diff
@@ -0,0 +1,96 @@
+# SUSE's openQA tests
+#
+# Copyright © 2020 SUSE LLC
+#
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved.  This file is offered as-is,
+# without any warranty.
+
+# Summary: Test Docker’s btrfs storage driver features for image and container management
+# Maintainer: qac team <qa-c@suse.de>
+
+package containers::runtime;
+use Mojo::Base -base;
+use testapi;
+use Test::Assert 'assert_equals';
+
+has runtime => undef;
+
+sub _rt_assert_script {
+    my ($self, $cmd, @args) = @_;
+    assert_script_run($self->runtime . " " . $cmd, @args);
+}
+
+sub _rt_script_run {
+    my ($self, $cmd, @args) = @_;
+    return script_run($self->runtime . " " . $cmd, @args);
+}
+
+sub _rt_script_output {
+    my ($self, $cmd, @args) = @_;
+    return script_output($self->runtime . " " . $cmd, @args);
+}
+
+sub build {
+    my ($self, $dockerfile_path, $container_tag) = @_;
+    die 'wrong number of arguments' if @_ < 3;
+    assert_script_run "cd $dockerfile_path";
+    $self->_rt_assert_script("build -t $container_tag .");
+    record_info "$container_tag created";
+}
+
+sub up {
+    my ($self, $tag, %args) = @_;
+    die 'tag is required' unless $tag;
+    my $mode   = $args{daemon} ? '-d'    : '-it';
+    my $remote = $args{cmd}    ? 'sh -c' : '';
+    my $ret    = $self->_rt_script_run(sprintf qq(run --rm %s %s %s '%s'), $mode, $tag, $remote, $args{cmd});
+    record_info "Remote run on $tag", $args{cmd};
+    return $ret;
+}
+
+sub pull {
+    my ($self, $tag) = @_;
+    my $ret = $self->_rt_script_run("pull $tag");
+    return $ret;
+}
+
+sub retrieve_images {
+    my ($self) = shift;
+    my $images_s = $self->_rt_script_output("images -q");
+    record_info "Images", $images_s;
+    my @images = split /[\n\t]/, $images_s;
+    return \@images;
+}
+
+sub retrieve_containers {
+    my ($self) = shift;
+    my $containers_s = $self->_rt_script_output("container ls -q");
+    record_info "Containers", $containers_s;
+    my @containers = split /[\n\t]/, $containers_s;
+    return \@containers;
+}
+
+sub info {
+    my ($self, %args) = shift;
+    my $property = $args{property} ? qq(--format '{{.$args{property}}}') : '';
+    my $expected = $args{value}    ? qq( | grep $args{value})            : '';
+    $self->_rt_assert_script(sprintf("info %s %s", $property, $expected));
+}
+
+sub remove_image {
+    my ($self, $tag) = @_;
+    $self->_rt_assert_script("rmi -f $tag");
+}
+
+sub cleanup_system {
```

Comment:
We have `containers::common::clean_container_host`.

Reply:
i am aware. but such function clearly belongs to the runtime. So yes it is overlap with the other until i will make a small refactoring.

Reply:
Can you please at least make a `# TODO ` comment then?

Reply:
if you want i can put a TODO but i have already a comment to remind me this anyway.

Reply:
> We have `containers::common::clean_container_host`.

and we need more clear name here . `clean_container_host` - clearly states what it does . `cleanup_system` - well I am not sure which `system` it is cleaning

Reply:
isnt it clear that it is about the host system that the docker/podman is running? if it is confusing lets address it in the follow up PR.

Reply:
no it is not , because you have host and guest and both are **systems** :)

Reply:
and all other functions in this class actually interact with guest

Reply:
clean_system_host it is

Reply:
I would say either `clean_host` or `clean_host_system` but I can probably fix it later..

Reply:
i am saving this for the follow up task. and i am resolving this for now.